pest/src/parser.rs

```diff
@@ -9,11 +9,25 @@
 
 use crate::error::Error;
 use crate::iterators::Pairs;
-use crate::RuleType;
+use crate::{RuleType, StateCheckpoint};
 
 /// A trait with a single method that parses strings.
-pub trait Parser<R: RuleType> {
+pub trait Parser<R: RuleType, S: Default + StateCheckpoint = ()> {
```

Comment:
won't this be a semver-breaking change? i.e. existing pest 2.X-generated code implemented the trait with one param and wouldn't compile with this/would need to be regenerated: https://github.com/pest-parser/pest/actions/runs/4276068348/jobs/7480923758#step:4:253

Reply:
Ugh, that's more breakage from the fact that the runtime and derive library versions aren't pinned to each other, isn't it... adding a new defaulted generic to a type is backwards compatible, as old code will use the default, but it's not forwards compatible, in that generated code supplying the parameter won't work with the older library versions without it.

Reply:
The bootstrap path will always be *interesting* because it has a weird mix of using the local version for some stuff and the crates-io version for other stuff.

Reply:
oh, yes, it's not forwards-compatible

Reply:
but that should be ok

Reply:
(sans figuring out the bootstrap)